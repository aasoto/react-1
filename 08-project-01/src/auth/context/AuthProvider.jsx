import { useReducer } from "react";
import { types } from "../types";
import { AuthContext } from "./AuthContext";
import { authReducer } from "./authReducer";

/** Funci贸n de inicializaci贸n */
const init = () => {
  /** Obtiene los datos del usuario en el localstorage */
  const user = JSON.parse(localStorage.getItem('user'));

  return {
    logged: !!user,
    user: user
  }
}

export const AuthProvider = ({children}) => {

  /** El authState es para obtener la informaci贸n que retorna el reducer.
   * El dispatch es para enviarle las acciones que este requiera.
   */
  const [authState, dispatch] = useReducer( authReducer, {}, init);

  const login = (name = '') => {

    const user = {id: 'ABC', name};

    const action = {
      type: types.login,
      payload: user
    }

    localStorage.setItem('user', JSON.stringify(user));

    dispatch(action);
  }

  const logout = () => {
    localStorage.removeItem('user');

    const action = {
      type: types.logout,
      payload: ''
    }

    dispatch(action);
  }

  return (
    <AuthContext.Provider value={{
      ...authState, //variables o datos que retornan del reducer de autenticaci贸n
      login,
      logout
    }}>
      {children}
    </AuthContext.Provider>
  )
}
