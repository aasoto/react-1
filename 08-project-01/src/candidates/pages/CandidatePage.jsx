import { useState } from "react";
import { useContext } from "react";
import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { SidebarContext } from "../../ui";
import { getCandidateById } from "../helpers";

export const CandidatePage = () => {
  
  const {showSidebar, pathUp} = useContext(SidebarContext);
  
  pathUp();

  const {id} = useParams();

  const navigate = useNavigate();

  const candidate = useMemo(() => getCandidateById(id), [id]);

  const { name, age, country, instagram, meet, placement } = candidate;
  
  const onNavigateBack = () => {
    navigate(-1);
  }

  if ( !candidate ) {
    return <Navigate to="/" />
  }

  return (
    <>
      <div className={`pt-24 pr-5 animate__animated ${ showSidebar ? 'pl-80 animate__fadeInLeft' : 'pl-5 animate__fadeInRight' }`}>
      <h1 className="text-gray-600 font-semibold text-3xl md:text-4xl mb-2">{name}</h1>
      <hr className="border"/>
      <div className={`mt-5 grid grid-cols-1 ${ showSidebar ? 'md:grid-cols-3' : 'md:grid-cols-4'} gap-5`}>
        <div className="col-span-1">
          <img src={`../images/candidates/${id}.PNG`} alt="" className="w-full object-cover object-top rounded-lg"/>
        </div>
        <div className={`${ showSidebar ? 'col-span-2' : 'col-span-3' }`}>
          <div className="grid grid-cols-3 gap-5">
            <div className="col-span-1">
              <span className={`fi fi-${id} scale-[6] md:scale-[10] translate-x-[53px] md:translate-x-[100px] translate-y-[40px] md:translate-y-[68px] rounded-sm`}></span>
            </div>
            <div className="col-span-2">
              <table className="w-full">
                <tbody>
                  <tr>
                    <td className="text-gray-500 font-medium text-lg">
                      Country:
                    </td>
                    <td colSpan={2} className="text-gray-600 font-bold text-2xl">
                      {country.toUpperCase()}
                    </td>
                  </tr>
                  <tr>
                    <td className="text-gray-500 font-medium text-lg">
                      Age:
                    </td>
                    <td colSpan={2} className="text-gray-600 font-bold text-2xl">
                      {age}
                    </td>
                  </tr>
                  <tr>
                    <td className="text-gray-500 font-medium text-lg">
                      Instagram:
                    </td>
                    <td colSpan={2} className="text-gray-600 text-xl">
                      <a href={`https://instagram.com/${instagram.slice(1)}`}>
                        {instagram}
                      </a>
                    </td>
                  </tr>
                </tbody>
              </table>  
            </div>
          </div>
          <h3 className="mt-20 text-gray-600 font-bold text-2xl">MEET</h3>
          <p className="text-gray-600 text-lg">{meet}</p>
        </div>
      </div>
      </div>
    </>
  )
}
