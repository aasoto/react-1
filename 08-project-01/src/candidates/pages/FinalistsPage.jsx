
export const FinalistsPage = () => {
  return (
    <div>FinalistsPage</div>
  )
}
