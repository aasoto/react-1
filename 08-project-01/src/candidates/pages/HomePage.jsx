import { CandidateList } from "../components"

export const HomePage = () => {
  return (
    <CandidateList page={'71th edition - Contestants'} placement={'all'}/>
  )
}
