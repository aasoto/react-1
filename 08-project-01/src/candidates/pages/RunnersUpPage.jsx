import { CandidateList } from "../components"

export const RunnersUpPage = () => {
  return (
    <CandidateList page={'Runners Up'} placement={'runners-up'}/>
  )
}
