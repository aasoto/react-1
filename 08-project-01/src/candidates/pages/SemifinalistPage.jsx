
export const SemifinalistPage = () => {
  return (
    <div>SemifinalistPage</div>
  )
}
