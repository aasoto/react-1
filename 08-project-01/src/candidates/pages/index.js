export * from './HomePage';
export * from './router';