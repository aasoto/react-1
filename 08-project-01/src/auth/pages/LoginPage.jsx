import React, { useContext } from 'react'
import { useEffect } from 'react';
import { useState } from 'react';
import { useForm } from '../../hooks'
import { AlertErrorAuth } from '../components';
import { AuthContext } from '../context';
import { authentication } from '../helpers';

export const LoginPage = () => {

  const {login} = useContext(AuthContext);

  const [alertErrorAuth, setAlertErrorAuth] = useState(false);
  
  const {email, password, onInputChange} = useForm({
    email: '',
    password: ''
  });

  useEffect(() => {
    if (alertErrorAuth) {
      setAlertErrorAuth(() => !alertErrorAuth);
    }
  }, [email, password]);

  const onLoginSubmit = (event) => {
    event.preventDefault();
    if ((email.trim().length <= 1) || (password.length <= 6)) return;

    const autenticated = authentication(email, password);

    if(autenticated) {
      login(autenticated.name);

    } else {
      setAlertErrorAuth(() => !alertErrorAuth);
    }
  }

  return (
    <>
      <div className='flex justify-center items-center h-screen'>
        <div className='w-72 md:w-96 h-80 bg-white rounded-lg shadow-xl'>
          <div className='w-full h-16 rounded-t-lg text-center border-b border-gray-400 flex justify-center items-center'>
            <h1 className='text-rose-700 text-3xl font-bold'>Login</h1>
          </div>
          <AlertErrorAuth show={alertErrorAuth}/>
          <form onSubmit={onLoginSubmit}>
            <div className='flex flex-col justify-center items-center gap-5 px-8 pt-10 pb-4'>
              <input 
                className='w-full h-12 px-5 border rounded-md border-gray-400 placeholder:text-gray-400' 
                type="email" 
                placeholder='Email'
                name="email"
                aria-label='email'
                value={email}
                onChange={onInputChange}
              />
              <input 
                className='w-full h-12 px-5 border rounded-md border-gray-400 placeholder:text-gray-400' 
                type="password" 
                placeholder='Password'
                name="password"
                aria-label='input-password'
                value={password}
                onChange={onInputChange}
              />
            </div>
            <div className='flex justify-end items-center gap-3 p-8'>
              <button className='bg-rose-700 hover:bg-rose-800 text-white font-bold rounded-md px-10 py-2 scale-100 hover:scale-105 transition duration-200'>Ok</button>
            </div>
          </form>
        </div>
      </div>
    </>
  )
}
