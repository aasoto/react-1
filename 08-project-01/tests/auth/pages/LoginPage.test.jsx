import { LoginPage } from "../../../src/auth/pages/LoginPage";
import { fireEvent, render, screen} from '@testing-library/react';
import { AuthContext } from "../../../src/auth/context";

const loginMock = jest.fn();

describe('Pruebas de <LoginPage />', () => {

  beforeEach(() => jest.clearAllMocks());

  test('debe hacer match con el snapshot', () => {

    const {container} = render(
      <AuthContext.Provider value={{login: loginMock}}>
        <LoginPage />
      </AuthContext.Provider>
    );

    expect(container).toMatchSnapshot();

  });

  test('no debe enviar los campos en blanco', () => {

    render(
      <AuthContext.Provider value={{login: loginMock}}>
        <LoginPage />
      </AuthContext.Provider>
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(loginMock).not.toHaveBeenCalled();

  });

  test('debe mostrar alerta de error cuando usuario y/o contraseña son incorrectos', () => {

    render(
      <AuthContext.Provider value={{login: loginMock}}>
        <LoginPage />
      </AuthContext.Provider>
    );

    const email = 'admin@gmail.com';
    const password = '123456789';

    const inputEmail = screen.getByRole('textbox', { name: 'email' });
    fireEvent.input(inputEmail, {target: {value: email}});
    expect(inputEmail.value).toBe(email);

    const inputPassword = screen.getByPlaceholderText('Password');
    fireEvent.input(inputPassword, {target: {value: password}});
    expect(inputPassword.value).toBe(password);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(screen.getByText('The email and/or password are wrong')).toBeTruthy();
    const alertErrorAuth = screen.getByLabelText('alert-error-auth');
    expect(alertErrorAuth.className).toContain('block');

  });

  test('debe llamar función login con argumentos', () => {
    
    render(
      <AuthContext.Provider value={{login: loginMock}}>
        <LoginPage />
      </AuthContext.Provider>
    );

    const email = 'admin@gmail.com';
    const password = '12345678';

    const inputEmail = screen.getByRole('textbox', { name: 'email' });
    fireEvent.input(inputEmail, {target: {value: email}});
    expect(inputEmail.value).toBe(email);

    const inputPassword = screen.getByPlaceholderText('Password');
    fireEvent.input(inputPassword, {target: {value: password}});
    expect(inputPassword.value).toBe(password);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(loginMock).toHaveBeenCalledWith("Administrador");
  });
});