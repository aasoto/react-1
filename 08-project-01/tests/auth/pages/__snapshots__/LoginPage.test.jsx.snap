// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <LoginPage /> debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="flex justify-center items-center h-screen"
  >
    <div
      class="w-72 md:w-96 h-80 bg-white rounded-lg shadow-xl"
    >
      <div
        class="w-full h-16 rounded-t-lg text-center border-b border-gray-400 flex justify-center items-center"
      >
        <h1
          class="text-rose-700 text-3xl font-bold"
        >
          Login
        </h1>
      </div>
      <div
        aria-label="alert-error-auth"
        class="w-full h-11 bg-red-200 border border-red-600 text-red-600 flex justify-center items-center hidden"
      >
        <h4>
          The email and/or password are wrong
        </h4>
      </div>
      <form>
        <div
          class="flex flex-col justify-center items-center gap-5 px-8 pt-10 pb-4"
        >
          <input
            aria-label="email"
            class="w-full h-12 px-5 border rounded-md border-gray-400 placeholder:text-gray-400"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
          <input
            aria-label="input-password"
            class="w-full h-12 px-5 border rounded-md border-gray-400 placeholder:text-gray-400"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="flex justify-end items-center gap-3 p-8"
        >
          <button
            class="bg-rose-700 hover:bg-rose-800 text-white font-bold rounded-md px-10 py-2 scale-100 hover:scale-105 transition duration-200"
          >
            Ok
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
